File: test-suite/src/test/java/io/micronaut/mqtt/docs/consumer/acknowledge/type/ProductListener.java
package io.micronaut.mqtt.docs.consumer.acknowledge.type;

// tag::imports[]
import io.micronaut.mqtt.annotation.Topic;
import io.micronaut.mqtt.annotation.MqttSubscriber;
import io.micronaut.messaging.Acknowledgement;

import java.util.concurrent.atomic.AtomicInteger;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "AcknowledgeSpec")
// tag::clazz[]
@MqttSubscriber
public class ProductListener {

    AtomicInteger messageCount = new AtomicInteger();

    @Topic("product")
    public void receive(byte[] data, Acknowledgement acknowledgement) { // <1>
        messageCount.getAndUpdate((intValue) -> ++intValue);
        acknowledgement.ack(); // <2>
    }
}
// end::clazz[]

File: test-suite-python/src/test/python/micronaut/mqtt/docs/consumer/acknowledge/type/ProductListener.py
# tag::imports[]
from micronaut.messaging import Acknowledgement
from micronaut.mqtt.annotation import MqttSubscriber, Topic
# end::imports[]

from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="AcknowledgeSpec")
# tag::clazz[]
@MqttSubscriber
class ProductListener:

    def __init__(self):
        self.message_count = 0

    @Topic("product")
    def receive(self, data: bytes, acknowledgement: Acknowledgement) -> None:  # <1>
        self.message_count += 1
        acknowledgement.ack()  # <2>
# end::clazz[]
